File: 53-maximum-subarray/maximum-subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        if (nums.length == 1) return nums[0];
        // int curr_sum = nums[0];
        int max_sum = nums[0], maxi = nums[0];
        for (int i = 1; i < nums.length; i++) {
            max_sum = Math.max(nums[i], max_sum + nums[i]);

            if (max_sum > maxi) {
                maxi = max_sum;
            }
        }


        return maxi;
    }
}

File: 53-maximum-subarray/maximum-subarray.py
class Solution:
    def maxSubArray(self, nums: list[int]) -> int:
        curr_sum = 0
        max_sum = nums[0]
        for n in nums:
            curr_sum += n
            max_sum = max(max_sum, curr_sum)

            if curr_sum < 0 :
                curr_sum = 0
        
        return max_sum